spring-cloud-gcp-pubsub/src/test/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubInboundChannelAdapterTests.java

```diff
@@ -31,13 +47,166 @@
 @RunWith(MockitoJUnitRunner.class)
 public class PubSubInboundChannelAdapterTests {
 
-	@Mock
-	private PubSubTemplate pubSubTemplate;
+	public static final String ACK = "ACK";
+
+	public static final String NACK = "NACK";
 
 	@Test(expected = IllegalArgumentException.class)
 	public void testNonNullAckMode() {
+		PubSubTemplate pubSubTemplate = Mockito.mock(PubSubTemplate.class);
+
 		PubSubInboundChannelAdapter adapter = new PubSubInboundChannelAdapter(
-				this.pubSubTemplate, "testSubscription");
+				pubSubTemplate, "testSubscription");
+
 		adapter.setAckMode(null);
 	}
+
+	@Test
+	public void testAckModeAuto() throws Exception {
+		PubSubTemplate pubSubTemplate = Mockito.mock(PubSubTemplate.class);
+		QueueChannel queueChannel = Mockito.mock(QueueChannel.class);
+
+		final Holder<String> returnValue = new Holder<>();
+		final CountDownLatch countDownLatch = new CountDownLatch(1);
+
+		PubSubInboundChannelAdapter adapter = new PubSubInboundChannelAdapter(
+				pubSubTemplate, "testSubscription");
+
+		adapter.setAckMode(AckMode.AUTO);
+		adapter.setSendTimeout(10);
+
+		Answer<MessageReceiver> answer = new Answer<MessageReceiver>() {
+			public MessageReceiver answer(InvocationOnMock invocation) throws Throwable {
+				Thread thread = new Thread(new Runnable() {
+					@Override
+					public void run() {
+						try {
+							PubsubMessage pubsubMessage = PubsubMessage.newBuilder().setData(
+									ByteString.copyFrom("Testing 1 2 3".getBytes("UTF-8"))).build();
+
+							Answer<Boolean> ackAnswer = new Answer<Boolean>() {
+								public Boolean answer(InvocationOnMock invocationOnMock) throws Throwable {
+									returnValue.value = ACK;
+									return null;
+								}
+							};
+
+							Answer<Boolean> nackAnswer = new Answer<Boolean>() {
+								public Boolean answer(InvocationOnMock invocationOnMock) throws Throwable {
+									returnValue.value = NACK;
+									return null;
+								}
+							};
+
+							AckReplyConsumer ackReplyConsumer = Mockito.mock(AckReplyConsumer.class);
```

Comment:
You can't stub in the active code.
you definitely have to do that in the main thread before any assignments.